transports/tcp/src/lib.rs

```diff
@@ -270,73 +312,140 @@ impl Future for TcpDialFut {
     /// Original configuration.
-    config: TcpConfig,
-    /// Generate `ListenerEvent::NewAddress` to inform about our listen address.
-    tell_new_addr: bool
+    config: TcpConfig
 }
 
 impl Stream for TcpListenStream {
     type Item = ListenerEvent<FutureResult<TcpTransStream, io::Error>>;
     type Error = io::Error;
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, io::Error> {
-        if self.tell_new_addr {
-            self.tell_new_addr = false;
-            return Ok(Async::Ready(Some(ListenerEvent::NewAddress(self.listen_addr.clone()))))
-        }
-
         let inner = match self.inner {
             Ok(ref mut inc) => inc,
-            Err(ref mut err) => {
-                return Err(err.take().expect("poll called again after error"));
-            }
+            Err(ref mut err) => return Err(err.take().expect("poll called again after error"))
         };
 
         loop {
-            match inner.poll() {
-                Ok(Async::Ready(Some(sock))) => {
-                    let addr = match sock.peer_addr() {
-                        // TODO: remove this expect()
-                        Ok(addr) => addr
-                            .to_multiaddr()
-                            .expect("generating a multiaddr from a socket addr never fails"),
-                        Err(err) => {
-                            // If we can't get the address of the newly-opened socket, there's
-                            // nothing we can do except ignore this connection attempt.
-                            error!("Ignored incoming because could't determine its \
-                                    address: {:?}", err);
-                            continue
-                        },
-                    };
-
-                    match apply_config(&self.config, &sock) {
-                        Ok(()) => (),
-                        Err(err) => {
-                            let event = ListenerEvent::Upgrade {
-                                upgrade: future::err(err),
-                                listen_addr: self.listen_addr.clone(),
-                                remote_addr: addr
-                            };
-                            return Ok(Async::Ready(Some(event)))
+            if let Some(event) = self.pending.pop_front() {
+                return Ok(Async::Ready(Some(event)))
+            }
+
+            let sock = match inner.poll() {
+                Ok(Async::Ready(Some(sock))) => sock,
+                Ok(Async::Ready(None)) => return Ok(Async::Ready(None)),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Err(()) => unreachable!("sleep_on_error never produces an error")
+            };
+
+            let sock_addr = match sock.peer_addr() {
+                Ok(addr) => addr,
+                Err(err) => {
+                    // If we can't get the address of the newly-opened socket, there's
```

Comment:
Actually, we should return `future::err` here.

Reply:
It is currently implemented this way on `master` but I agree that returning an error is more appropriate.

Reply:
When you wrote `future::err` you meant [`Err(io::Error ...)`](https://github.com/libp2p/rust-libp2p/pull/1044/commits/da55db23b8f6b18f7f78751f8fef48bb55fb057b), right?

Reply:
I was thinking of returning `Ok(future::err(...))`, similar to lines 450-454.

Reply:
This would require a change to `ListenerEvent` since we have no value for `remote_addr`.

Reply:
I also think that an error from `TcpStream::peer_addr` and `TcpListener::local_addr` within this loop would be serious problem. Why should it fail at all? On POSIX systems for example, `peer_addr` maps to libc's `getpeername` and `local_addr` to `getsockname`. Here is the union of errors those functions return:
```
EBADF    The argument sockfd is not a valid file descriptor.
EFAULT   The addr argument points to memory not in a valid part of the process address space.
EINVAL   addrlen is invalid (e.g., is negative).
ENOBUFS  Insufficient resources were available in the system to perform the operation.
ENOTCONN The socket is not connected.
ENOTSOCK The file descriptor sockfd does not refer to a socket.
```
None of this should happen, but if it does I think we are in trouble and a `return Err(...)` is warranted.